In [6]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

In [7]:
!pip install plotly
import plotly.graph_objects as go
def paint(train_losses, train_r2_scores, test_losses, test_r2_scores):
  # Plot training loss
  loss_fig = go.Figure()
  loss_fig.add_trace(go.Scatter(x=list(range(1, num_epochs + 1)), y=train_losses, mode='lines', name='Training Loss'))
  loss_fig.update_layout(title='Training Loss over Epochs', xaxis_title='Epoch', yaxis_title='Loss')


  # Plot training R2 score
  r2_fig = go.Figure()
  r2_fig.add_trace(go.Scatter(x=list(range(1, num_epochs + 1)), y=train_r2_scores, mode='lines', name='Training R2 Score'))
  r2_fig.update_layout(title='Training R2 Score over Epochs', xaxis_title='Epoch', yaxis_title='R2 Score')
  loss_fig.show()
  r2_fig.show()
  # Plot training loss
  loss_fig = go.Figure()
  loss_fig.add_trace(go.Scatter(x=list(range(1, num_epochs + 1)), y=test_losses, mode='lines', name='Test Loss'))
  loss_fig.update_layout(title='Test Loss over Epochs', xaxis_title='Epoch', yaxis_title='Loss')


  # Plot training R2 score
  r2_fig = go.Figure()
  r2_fig.add_trace(go.Scatter(x=list(range(1, num_epochs + 1)), y=test_r2_scores, mode='lines', name='Test R2 Score'))
  r2_fig.update_layout(title='Test R2 Score over Epochs', xaxis_title='Epoch', yaxis_title='R2 Score')
  loss_fig.show()
  r2_fig.show()


In [9]:
#from google.colab import drive
#drive.mount('/content/drive')
anger_train = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/train/anger-train.csv')
anger_test = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/test/anger-test.csv')

fear_train = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/train/fear-train.csv')
fear_test = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/test/fear-test.csv')

joy_train = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/train/joy-train.csv')
joy_test = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/test/joy-test.csv')

sadness_train = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/train/sadness-train.csv')
sadness_test = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/EI/test/sadness-test.csv')

V_train = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/V/Valence-train.csv')
V_test = pd.read_csv('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/V/Valence-test.csv')

In [10]:
anger_train.head()

Score                                               Text
0  0.562              "shut up hashtags are cool #offended"
1  0.750  "it makes me so fucking irate jesus. nobody is...
2  0.417       "Lol Adam the Bull with his fake outrage..."
3  0.354  "passed away early this morning in a fast and ...
4  0.438  "lol wow i was gonna say really?! haha have yo...

In [11]:
anger_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1701 entries, 0 to 1700
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Score   1701 non-null   float64
 1   Text    1701 non-null   object 
dtypes: float64(1), object(1)
memory usage: 26.7+ KB


In [12]:
##ну типо закинуть вв вмодель попробовать обучить на слова и несколько обученных моделий под каждую эмоцию и валидацию adv и эту сравнить
#злость
X_anger_train = pd.DataFrame(anger_train['Text'])
X_anger_test = pd.DataFrame(anger_test['Text'])

Y_anger_train = anger_train['Score'].tolist()
Y_anger_test = anger_test['Score'].tolist()
#страх
X_fear_train = pd.DataFrame(fear_train['Text'])
X_fear_test = pd.DataFrame(fear_test['Text'])

Y_fear_train = fear_train['Score']
Y_fear_test = fear_test['Score']
#наслождение
X_joy_train = pd.DataFrame(joy_train['Text'])
X_joy_test = pd.DataFrame(joy_test['Text'])

Y_joy_train = joy_train['Score']
Y_joy_test = joy_test['Score']
#грусть
X_sadness_train = pd.DataFrame(sadness_train['Text'])
X_sadness_test = pd.DataFrame(sadness_test['Text'])

Y_sadness_train = sadness_train['Score']
Y_sadness_test = sadness_test['Score']
#валентность))) (плохой-хороший)
X_V_train = pd.DataFrame(V_train['Text'])
X_V_test = pd.DataFrame(V_test['Text'])

Y_V_train = V_train['Score']
Y_V_test = V_test['Score']

In [13]:
X_anger_train.head()

Text
0              "shut up hashtags are cool #offended"
1  "it makes me so fucking irate jesus. nobody is...
2       "Lol Adam the Bull with his fake outrage..."
3  "passed away early this morning in a fast and ...
4  "lol wow i was gonna say really?! haha have yo...

In [14]:
X_anger_test.head()

Text
0    "I know you mean well but I'm offended. Prick."
1  "Let go of resentment, it will hold you back, ...
2  "No, I'm not 'depressed because of the weather...
3  "#AmarnathTerrorAttack Muslims are killing eve...
4  "Prepare to suffer the sting of Ghost Rider's ...

In [15]:

import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize
nltk.download('stopwords')
from nltk.corpus import stopwords
stop_words = stopwords.words("english")
def remove_stopwords(text):
    tokens = word_tokenize(text)
    filtered_tokens = [token for token in tokens if token.lower() not in stop_words]
    filtered_text = ' '.join(filtered_tokens)
    return filtered_text
import re
def clean_data(df):
    # Remove HTML tags
    df['Text'] = df['Text'].apply(lambda x: re.sub('<.*?>', '', x))

    # Remove non-alphanumeric characters and convert to lowercase
    df['Text'] = df['Text'].apply(lambda x: re.sub('[^0-9a-zA-Z\s]+', '', x).lower())
    df['Text'] = df['Text'].apply(remove_stopwords)

    return df

[nltk_data] Downloading package punkt to /Users/kidawr/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/kidawr/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [16]:
X_anger_train = clean_data(X_anger_train)
X_anger_test = clean_data(X_anger_test)

X_fear_train = clean_data(X_fear_train)
X_fear_test = clean_data(X_fear_test)

X_joy_train = clean_data(X_joy_train)
X_joy_test = clean_data(X_joy_test)

X_sadness_train = clean_data(X_sadness_train)
X_sadness_test = clean_data(X_sadness_test)

X_V_train = clean_data(X_V_train)
X_V_test = clean_data(X_V_test)

In [17]:
X_anger_train = X_anger_train['Text'].tolist()
X_anger_test = X_anger_test['Text'].tolist()

X_fear_train = X_fear_train['Text'].tolist()
X_fear_test = X_fear_test['Text'].tolist()

X_joy_train = X_joy_train['Text'].tolist()
X_joy_test = X_joy_test['Text'].tolist()

X_sadness_train = X_sadness_train['Text'].tolist()
X_sadness_test = X_sadness_test['Text'].tolist()

X_V_train = X_V_train['Text'].tolist()
X_V_test = X_V_test['Text'].tolist()

In [18]:
#токенезируем наш текст с помощью трансформера bert
!pip install transformers
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

Xb_anger_train = tokenizer(X_anger_train, truncation=True, padding=True)
Xb_anger_test = tokenizer(X_anger_test, truncation=True, padding=True)

Xb_fear_train = tokenizer(X_fear_train, truncation=True, padding=True)
Xb_fear_test = tokenizer(X_fear_test, truncation=True, padding=True)

Xb_joy_train = tokenizer(X_joy_train, truncation=True, padding=True)
Xb_joy_test = tokenizer(X_joy_test, truncation=True, padding=True)

Xb_sadness_train = tokenizer(X_sadness_train, truncation=True, padding=True)
Xb_sadness_test = tokenizer(X_sadness_test, truncation=True, padding=True)

Xb_V_train = tokenizer(X_V_train, truncation=True, padding=True)
Xb_V_test = tokenizer(X_V_test, truncation=True, padding=True)

Downloading:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

###  Train the Model

In [19]:
import torch
    
class ValueDataset(torch.utils.data.Dataset):
    def __init__(self, encodings,values):
        self.encodings = encodings
        self.values = list(values)
        #Convert each value to float
    
    def __len__(self):
        return len(self.values)
    
    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['values'] = torch.tensor(self.values[idx])
        return item



anger_train_dataset = ValueDataset(Xb_anger_train, Y_anger_train)
anger_test_dataset = ValueDataset(Xb_anger_test, Y_anger_test)

fear_train_dataset = ValueDataset(Xb_fear_train, Y_fear_train)
fear_test_dataset = ValueDataset(Xb_fear_test, Y_fear_test)

joy_train_dataset = ValueDataset(Xb_joy_train, Y_joy_train)
joy_test_dataset = ValueDataset(Xb_joy_test, Y_joy_test)

sadness_train_dataset = ValueDataset(Xb_sadness_train, Y_sadness_train)
sadness_test_dataset = ValueDataset(Xb_sadness_test, Y_sadness_test)

V_train_dataset = ValueDataset(Xb_V_train, Y_V_train)
V_test_dataset = ValueDataset(Xb_V_test, Y_V_test)

In [20]:
from transformers import BertTokenizer
from torch.utils.data import DataLoader  # Add this line

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')


train_loader_array = [DataLoader(anger_train_dataset, batch_size=16, shuffle=True), 
                      DataLoader(fear_train_dataset, batch_size=16, shuffle=True), 
                      DataLoader(joy_train_dataset, batch_size=16, shuffle=True), 
                      DataLoader(sadness_train_dataset, batch_size=16, shuffle=True), 
                      DataLoader(V_train_dataset, batch_size=16, shuffle=True)]

test_loader_array = [DataLoader(anger_test_dataset, batch_size=16, shuffle=True), 
                     DataLoader(fear_test_dataset, batch_size=16, shuffle=True), 
                     DataLoader(joy_test_dataset, batch_size=16, shuffle=True), 
                     DataLoader(sadness_test_dataset, batch_size=16, shuffle=True), 
                     DataLoader(V_test_dataset, batch_size=16, shuffle=True)]


In [25]:
import torch
import copy

def Model_A(model):
    #################
    # Загрузка весов модели
    model_A = copy.deepcopy(model)
    model_A.load_state_dict(torch.load('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_A.pth', map_location=torch.device('cpu')))

    # Перевод модели в режим оценки (eval mode)
    model_A.eval()
    ##########
    return model_A

def Model_F(model):
    #################
    # Загрузка весов модели
    model_F = copy.deepcopy(model)
    model_F.load_state_dict(torch.load('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_F.pth', map_location=torch.device('cpu')))

    # Перевод модели в режим оценки (eval mode)
    model_F.eval()
    ##########
    return model_F

def Model_J(model):
    #################
    # Загрузка весов модели
    model_J = copy.deepcopy(model)
    model_J.load_state_dict(torch.load('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_J.pth', map_location=torch.device('cpu')))

    # Перевод модели в режим оценки (eval mode)
    model_J.eval()
    ##########
    return model_J

def Model_S(model):
    #################
    # Загрузка весов модели
    model_S = copy.deepcopy(model)
    model_S.load_state_dict(torch.load('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_S.pth', map_location=torch.device('cpu')))

    # Перевод модели в режим оценки (eval mode)
    model_S.eval()
    return model_S

def Model_V(model):
    #################
    # Загрузка весов модели
    model_V = copy.deepcopy(model)
    model_V.load_state_dict(torch.load('/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_V.pth', map_location=torch.device('cpu')))

    # Перевод модели в режим оценки (eval mode)
    model_V.eval()
    ##########
    return model_V


In [26]:
flag = True # True-загрузить веса, False-обучить модель, после обучения надо загружать сохраненые веса
import torch
torch.cuda.empty_cache()
import copy
# Check current GPU memory usage
allocated_memory = torch.cuda.memory_allocated()
print(f"Current GPU memory allocated: {allocated_memory / 1024**3} GB")
import torch
from torch.nn import MSELoss
from transformers import BertForSequenceClassification, AdamW
from torch.utils.data import DataLoader
from sklearn.metrics import r2_score

# Define the number of epochs
num_epochs = 50

# Use GPU if available, else use CPU
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
Model_Count = 0# в зависимости от значения будем сохранят и загружать веса разных моделий
model_A,model_F,model_J,model_S,model_V = None,None,None,None,None
for train_loader, test_loader in zip(train_loader_array, test_loader_array):
  Model_Count += 1
  # Load pre-trained model
  model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=1)
  model.to(device)

  # Set up the optimizer
  optimizer = AdamW(model.parameters(), lr=1e-5)

  # Define the loss function
  loss_fn = MSELoss()

  def evaluate(model, data_loader, loss_fn, device):
      model.eval()
      r2_scores = []
      losses = []
      with torch.no_grad():
          for batch in data_loader:
              input_ids = batch["input_ids"].to(device)
              attention_mask = batch["attention_mask"].to(device)
              targets = batch["values"].to(device)

              outputs = model(input_ids=input_ids, attention_mask=attention_mask)
              preds = outputs.logits.squeeze()
              #функция r2_score требует наличия нескольких точек для корректного вычисления коэффициента детерминации R2 если 1 то игнорируем
              if len(targets) > 1 and len(preds) > 1:
                r2 = r2_score(targets.detach().cpu().numpy(), preds.detach().cpu().numpy())
                r2_scores.append(r2)
                loss = loss_fn(preds, targets)
                losses.append(loss.item())

      return np.mean(r2_scores), np.mean(losses)


  def train_epoch(model, data_loader, loss_fn, optimizer, device):
      model.train()
      losses = []
      r2_scores = []
      for batch_idx, d in enumerate(data_loader, 1):
          input_ids = d["input_ids"].to(device)
          attention_mask = d["attention_mask"].to(device)
          targets = d["values"].to(device)

          optimizer.zero_grad()
          outputs = model(input_ids=input_ids, attention_mask=attention_mask)
          preds = outputs.logits.squeeze()
          loss = loss_fn(preds, targets)
          loss.backward()
          optimizer.step()

          losses.append(loss.item())
          r2 = r2_score(targets.detach().cpu().numpy(), preds.detach().cpu().numpy())
          r2_scores.append(r2)

          print(f"\rEpoch {epoch}/{num_epochs} {batch_idx}/{len(data_loader)} "
                f"- loss: {loss.item():.4f} - r2_score: {r2:.4f}", end="")

      return losses, r2_scores
  if flag:
    if Model_Count == 1:
      model_A = Model_A(model)
    elif Model_Count == 2:
      model_F = Model_F(model)
    elif Model_Count == 3:
      model_J = Model_J(model)
    elif Model_Count == 4:
      model_S = Model_S(model)
    elif Model_Count == 5:
      model_V = Model_V(model)
  else:
    train_losses = []
    train_r2_scores = []
    test_r2_scores =[]
    test_losses = []
    for epoch in range(1, num_epochs + 1):
        print(f"\nEpoch {epoch}/{num_epochs}")

        train_loss, train_r2_score = train_epoch(model, train_loader, loss_fn, optimizer, device)
        train_losses.extend(train_loss)
        train_r2_scores.extend(train_r2_score)
        print(f" Train loss: {np.mean(train_losses):.4f} - Train R2 score: {np.mean(train_r2_scores):.4f}")
        test_r2_score, test_loss = evaluate(model, test_loader, loss_fn, device)
        test_r2_scores.append(test_r2_score)
        test_losses.append(test_loss)
        print(f"Test R2 score: {test_r2_score:.4f} Test loss: {test_loss:.4f}")

    print(f"\nFinal Train loss: {np.mean(train_losses):.4f}")
    #Рисовашки
    paint(train_losses, train_r2_scores, test_losses, test_r2_scores)
    #СОХРАНЕНИЕ ВЕСОВ
    Dict = {
        1: lambda: torch.save(model.state_dict(), '/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_A.pth'),
        2: lambda: torch.save(model.state_dict(), '/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_F.pth'),
        3: lambda: torch.save(model.state_dict(), '/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_J.pth'),
        4: lambda: torch.save(model.state_dict(), '/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_S.pth'),
        5: lambda: torch.save(model.state_dict(), '/Users/kidawr/Desktop/Сириус/Высокоуровневые языки/concordia-ml/semeval/Wheids/model_weights_V.pth')
    }
    Dict[Model_Count]()  # сохраняем веса

Current GPU memory allocated: 0.0 GB


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.bias']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at

In [17]:
#@title Текст заголовка по умолчанию воу воу обучается))


In [30]:
# чистим от мусора предложение
def clean_words(sentence):
    # Удаление HTML-тегов
    sentence = re.sub('<.*?>', '', sentence)

    # Удаление неалфавитно-цифровых символов и приведение к нижнему регистру
    sentence = re.sub('[^0-9a-zA-Z\s]+', '', sentence).lower()
    tokens = word_tokenize(sentence)
    filtered_tokens = [token for token in tokens if token.lower() not in stop_words]
    sentence = ' '.join(filtered_tokens)

    # токикизируем с бертой)
    inputs = tokenizer(sentence, truncation=True, padding=True, return_tensors="pt")
    # Передача входных данных в модель
    input_ids = inputs["input_ids"].to(device)
    attention_mask = inputs["attention_mask"].to(device)
    return input_ids, attention_mask

def Print_res(sentence):
    input_ids, attention_mask = clean_words(sentence)
    # Получение оценки предложения
    outputs = model_A(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    # Оценка предсказания
    prediction = preds.item()
    print(f"angry: {prediction}")

    outputs = model_F(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    prediction = preds.item()
    print(f"fear: {prediction}")


    outputs = model_J(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    prediction = preds.item()
    print(f"joy: {prediction}")

    outputs = model_S(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    prediction = preds.item()
    print(f"sadness: {prediction}")

    outputs = model_V(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    prediction = preds.item()
    print(f"valent: {prediction}")
Print_res("If you ignore me, you will die!")

angry: 0.4226667881011963
fear: 0.5345861315727234
joy: 0.2506936490535736
sadness: 0.5730095505714417
valent: 0.39034920930862427


In [ ]:
sum = 0
delts= []
predictions = []
#anger_test fear_test joy_test sadness_test V_test
namedt = "fear_test"
dt = fear_test
model = model_F
for sentence, score in zip(dt['Text'], dt['Score']):
    input_ids, attention_mask = clean_words(sentence)
    outputs = model(input_ids=input_ids, attention_mask=attention_mask)
    preds = outputs.logits.squeeze()
    # Оценка предсказания
    prediction = preds.item()
    delts.append(abs(score - prediction))
    predictions.append(prediction)

# Запись предсказаний в новый столбец "Test_score"
dt['Test_score'] = predictions

print(f"mean ошбка: {np.mean(delts)}")
print(f"max ошибка на тестовой: {np.max(delts)}")
# Сохранение данных в новый файл
dt.to_csv('/content/drive/MyDrive/python/semeval/Test_score/' + namedt + '.csv', index=False)

mean ошбка: 0.10085645992818153
max ошибка на тестовой: 0.498538910150528


A-злость:

mean ошбка: 0.11225102423788783

max ошибка на тестовой: 0.45749985551834105

F-страх:

mean ошбка: 0.10085645992818153

max ошибка на тестовой: 0.498538910150528

J-наслождение:

mean ошбка: 0.12240305352286518

max ошибка на тестовой: 0.6921054396629334

S-грусть:

mean ошбка: 0.11097818684626848

max ошибка на тестовой: 0.5690200281143188

V-0 плохо,1 хорошо:

mean ошбка: 0.1059550210799835

max ошибка на тестовой: 0.4936003503799439